File: tesi/realtesi/Temporal_SM/src/main/scala/FileManagerTemporal.scala
import java.io.BufferedReader

import java.io.FileReader

import java.io.BufferedWriter

import java.io.FileWriter

import java.util.Vector

import gnu.trove.iterator.TIntIterator

import gnu.trove.set.hash.TIntHashSet


class FileManagerTemporal {

  /*
	Read a graph from input file
	@param graphFile: path of the file
	*/

  def readGraph(graphFile: String): TemporalGraph = {
    var g: TemporalGraph = null
    try {
      val br: BufferedReader = new BufferedReader(new FileReader(graphFile))
      var str: String = br.readLine()
      var directed: Boolean = false
      if (str.==("directed")) directed = true
      val numNodes: Int = java.lang.Integer.parseInt(br.readLine())
      g = new TemporalGraph(directed, numNodes)
      str = br.readLine()
      while (str!=null) {
        val split: Array[String] = str.split("[, \t]")
        val source: Int = java.lang.Integer.parseInt(split(0))
        val dest: Int = java.lang.Integer.parseInt(split(1))
        val time: Int = java.lang.Integer.parseInt(split(2))
        g.addEdge(source, dest,time)
        str = br.readLine()
      }
      br.close()
    } catch {
      case e: Exception => println(e.getMessage)

    }
    g
  }

  /*
	Read the set of queries from input file
	@param queriesFile: path of the file
	*/

  def readQueries(queriesFile: String): Vector[TemporalGraph] = {
    val setQueries: Vector[TemporalGraph] = new Vector[TemporalGraph]()
    try {
      val br: BufferedReader = new BufferedReader(new FileReader(queriesFile))
      var str: String = br.readLine()
      while (str != null) {
        val direction: String = br.readLine()
        var directed: Boolean = false
        if (direction.==("directed")) directed = true
        val numNodes: Int = java.lang.Integer.parseInt(br.readLine())
        val q: TemporalGraph = new TemporalGraph(directed, numNodes)
        str = br.readLine()
        while (str  != null && !str.startsWith("#")) {
          val split: Array[String] = str.split("[ \t]")
          val source: Int = java.lang.Integer.parseInt(split(0))
          val dest: Int = java.lang.Integer.parseInt(split(1))
          val time: Int = java.lang.Integer.parseInt(split(2))
          q.addEdge(source, dest,time)
          str = br.readLine()
        }
        setQueries.add(q)
      }
      br.close()
    } catch {
      case e: Exception => println(e.toString)

    }
    setQueries
  }

  /*
	Write query results to output file
	*/

  def writeResults(setQueries: Vector[TemporalGraph],
                   setCounts: Vector[Long],
                   setRunningTimes: Vector[Double],
                   outputFile: String): Unit = {
    try {
      val bw: BufferedWriter = new BufferedWriter(new FileWriter(outputFile))
      bw.write("Query\tNum_occ\tTime (secs)\n")
      for (i <- 0 until setQueries.size) {
        val q: TemporalGraph = setQueries.get(i)
        val adjString: String = q.getAdjString
        bw.write(
          adjString + "\t" + setCounts.get(i) + "\t" + setRunningTimes.get(i) +
            "\n")
      }
      bw.close()
    } catch {
      case e: Exception => println(e.getMessage)

    }
  }

}

File: tesi/realtesi/Java_Temporal/RITemporal/src/FileManagerTemporal.java
/*
Class for handling input and output operations
*/

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.util.HashMap;
import java.util.Vector;

public class FileManagerTemporal
{
	
	int maxNodeMap=0;
	
	HashMap<Integer,NameMapping> mappingNames= new HashMap<>();
    
	/*
	Read a graph from input file
	@param graphFile: path of the file
	*/
	public TemporalGraph readGraph(String graphFile)
    {
        TemporalGraph g=null;
        try
        {
            BufferedReader br=new BufferedReader(new FileReader(graphFile));
            String str=br.readLine();
            boolean directed=false;
            if(str.equals("directed"))
                directed=true;
            int numNodes=Integer.parseInt(br.readLine());
            int source = -1;
            int destination = -1;
            g=new TemporalGraph(directed,numNodes);
            while((str=br.readLine())!=null)
            {
                String[] split=str.split("[, \t]");
                String sourcestr=split[0];
                String deststr=split[1];
                if(!mappingNames.containsKey(sourcestr.hashCode())) {
                	mappingNames.put(sourcestr.hashCode(), new NameMapping(maxNodeMap, sourcestr));
                	source = maxNodeMap;
                	maxNodeMap++;
                }else {
                	source = mappingNames.get(sourcestr.hashCode()).map;
                }
                
                if(!mappingNames.containsKey(deststr.hashCode())) {
                	mappingNames.put(deststr.hashCode(), new NameMapping(maxNodeMap, deststr));
                	destination = maxNodeMap;
                	maxNodeMap++;
                }else {
                	destination = mappingNames.get(deststr.hashCode()).map;
                }
                int time=Integer.parseInt(split[2]);
                g.addEdge(source,destination,time);
            }
            br.close();
        }
        catch(Exception e){
            System.out.println(e.getMessage());
        }
        return g;
    }

    /*
	Read the set of queries from input file
	@param queriesFile: path of the file
	*/
    public Vector<TemporalGraph> readQueries(String queriesFile)
    {
        Vector<TemporalGraph> setQueries=new Vector<>();
        try
        {
            BufferedReader br=new BufferedReader(new FileReader(queriesFile));
            String str=br.readLine();
            while(str!=null)
            {
                String direction=br.readLine();
                boolean directed=false;
                if(direction.equals("directed"))
                    directed=true;
                int numNodes=Integer.parseInt(br.readLine());
                TemporalGraph q=new TemporalGraph(directed,numNodes);
                while((str=br.readLine())!=null && !str.startsWith("#"))
                {
                    String[] split=str.split("[ \t]");
                    int source=Integer.parseInt(split[0]);
                    int dest=Integer.parseInt(split[1]);
                    int time=Integer.parseInt(split[2]);
                    q.addEdge(source,dest,time);
                }
                setQueries.add(q);
            }
            br.close();
        }
        catch(Exception e){
            System.out.println(e.toString());
        }
        return setQueries;
    }

	/*
	Write query results to output file
	*/
    public void writeResults(Vector<Graph> setQueries, Vector<Long> setCounts, Vector<Double> setRunningTimes, String outputFile)
    {
        try
        {
            BufferedWriter bw = new BufferedWriter(new FileWriter(outputFile));
            bw.write("Query\tNum_occ\tTime (secs)\n");
            int i;
            for(i=0;i<setQueries.size();i++)
            {
                Graph q=setQueries.get(i);
                String adjString=q.getAdjString();
                bw.write(adjString+"\t"+setCounts.get(i)+"\t"+setRunningTimes.get(i)+"\n");
            }
            bw.close();
        }
        catch(Exception e){
            System.out.println(e.getMessage());
        }
    }

}
